### PR TITLE
feat: add setServerUrl api

### PR DESCRIPTION
**Pull request by yuhao900914**
### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Java/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

**Review by haoliu-amp — CHANGES_REQUESTED** [review]
User might pass the EU batch endpoint. It also ends with `batch` but it is `https://api.eu.amplitude.com/batch`

Refer to https://developers.amplitude.com/docs/batch-event-upload-api

With current logic, I believe it will be treated as a normal endpoint. 

cc: @qingzhuozhen

**Comment by yuhao900914** [comment]
> User might pass the EU batch endpoint. It also ends with `batch` but it is `https://api.eu.amplitude.com/batch`
> 
> Refer to https://developers.amplitude.com/docs/batch-event-upload-api
> 
> With current logic, I believe it will be treated as a normal endpoint.
> 
> cc: @qingzhuozhen

@haoliu-amp 
`setServerUrl ` only used for setting amplitude endpoints, right? 
Can I use the `batch` keyword in the endpoint to determine if it's a regular request or batch request?

**Review by haoliu-amp — CHANGES_REQUESTED** [review]
Look good to me. Could we write a fews tests for sending data to EU API and EU BATCH API?

File: src/main/java/com/amplitude/Amplitude.java [hunk 1]
```diff
@@ -81,12 +95,12 @@ public void useBatchMode(Boolean isBatchMode) {
...
-      httpCallMode = updatedHttpCallMode;
+    httpCallMode = updatedHttpCallMode;
+    if (updatedHttpCallMode == HttpCallMode.BATCH_HTTPCALL) {
+      httpCall = new BatchHttpCall(apiKey, serverUrl != null ? serverUrl : Constants.BATCH_API_URL);
+    } else {
       httpCall =
-          (updatedHttpCallMode == HttpCallMode.BATCH_HTTPCALL)
-              ? new BatchHttpCall(apiKey)
-              : new GeneralHttpCall(apiKey);
+          new GeneralHttpCall(apiKey, serverUrl != null ? serverUrl : Constants.BATCH_API_URL);
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/Amplitude.java:103]: The default for general is not batch url right?

File: src/main/java/com/amplitude/BatchHttpCall.java [hunk 2]
```diff
@@ -1,10 +1,11 @@
 package com.amplitude;
 
 public class BatchHttpCall extends HttpCall {
-  private static String apiUrl = Constants.BATCH_API_URL;
+  private static String apiUrl;
 
-  protected BatchHttpCall(String apiKey) {
+  protected BatchHttpCall(String apiKey, String batchServerUrl) {
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/BatchHttpCall.java:6]: This is ok, we can also adjust the `HttpCall` constructor and set from there.

File: src/main/java/com/amplitude/BatchHttpCall.java [hunk 3]
```diff
@@ -1,10 +1,11 @@
 package com.amplitude;
 
 public class BatchHttpCall extends HttpCall {
-  private static String apiUrl = Constants.BATCH_API_URL;
+  private static String apiUrl;
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/BatchHttpCall.java:4]: why we make this static earlier? I don't see we have a strong need for it.

File: src/main/java/com/amplitude/Amplitude.java (hunk 1 above)
Comment (yuhao900914) [reply src/main/java/com/amplitude/Amplitude.java:103]: Yes, the default is the general one.

File: src/main/java/com/amplitude/Amplitude.java [hunk 4]
```diff
@@ -70,6 +71,19 @@ public void init(String key) {
     updateHttpCall(HttpCallMode.REGULAR_HTTPCALL);
   }
 
+  /**
+   * Sends events to a different URL other than Constants.API_URL or Constants.BATCH_API_URL. Used
+   * for proxy server.
+   *
+   * @param url the server URL for sending event
+   */
+  public void setServerUrl(String url) {
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/Amplitude.java:80]: let's add some test case for it.

File: src/main/java/com/amplitude/Amplitude.java (hunk 1 above)
Comment (qingzhuozhen) [reply src/main/java/com/amplitude/Amplitude.java:103]: I mean when serverUrl is null here the fallback url is batch url not the general one.

File: src/main/java/com/amplitude/BatchHttpCall.java (hunk 3 above)
Comment (haoliu-amp) [reply src/main/java/com/amplitude/BatchHttpCall.java:4]: +1

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: src/main/java/com/amplitude/Amplitude.java (hunk 1 above)
Comment (yuhao900914) [reply src/main/java/com/amplitude/Amplitude.java:103]: Oh, good catch! Will fix this.

File: src/test/java/com/amplitude/AmplitudeTest.java [hunk 5]
```diff
@@ -122,6 +124,34 @@ public void testLogEventWithInvalidResponse(boolean useBatch)
...
+  public void testDefaultServerUrl(boolean useBatch)
+      throws NoSuchFieldException, IllegalAccessException {
+    Amplitude amplitude = Amplitude.getInstance("test");
+    amplitude.init(apiKey);
+    amplitude.useBatchMode(useBatch);
+
+    Field httpCallField = amplitude.getClass().getDeclaredField("httpCall");
+    httpCallField.setAccessible(true);
+    HttpCall httpCall = (HttpCall) httpCallField.get(amplitude);
+    System.out.println(httpCall.getApiUrl());
```
Comment (qingzhuozhen) [inline src/test/java/com/amplitude/AmplitudeTest.java:138]: clean up this?

Comment (yuhao900914) [reply src/test/java/com/amplitude/AmplitudeTest.java]: Cleaned.

**Review by qingzhuozhen — APPROVED** [review]

File: src/test/java/com/amplitude/AmplitudeTest.java [hunk 6]
```diff
@@ -122,6 +124,33 @@ public void testLogEventWithInvalidResponse(boolean useBatch)
...
+    amplitude.useBatchMode(useBatch);
+
+    Field httpCallField = amplitude.getClass().getDeclaredField("httpCall");
+    httpCallField.setAccessible(true);
+    HttpCall httpCall = (HttpCall) httpCallField.get(amplitude);
+    assertEquals(httpCall.getApiUrl(), useBatch ? Constants.BATCH_API_URL : Constants.API_URL);
+  }
+
+  @Test
+  public void testSetServerUrl() throws NoSuchFieldException, IllegalAccessException {
```
Comment (haoliu-amp) [inline src/test/java/com/amplitude/AmplitudeTest.java:142]: Let's add a test for EU batch API

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

**Review by haoliu-amp — APPROVED** [review]

**Pull request merged**